### PR TITLE
PROTON-2417 Add errorprone to the maven build

### PR DESCRIPTION
**Pull request by jiridanek**

### CODE REVIEW

**Comment by jiridanek** [comment]
Few examples of what it reports, from https://github.com/apache/qpid-protonj2/runs/3412406596

```
Warning:  /home/runner/work/qpid-protonj2/qpid-protonj2/protonj2-test-driver/src/main/java/org/apache/qpid/protonj2/test/driver/codec/transport/Attach.java:[273,16] [FallThrough] Execution may fall through from the previous case; add a `// fall through` comment before this line if it was deliberate
    (see https://errorprone.info/bugpattern/FallThrough)
```

```
Warning:  /home/runner/work/qpid-protonj2/qpid-protonj2/protonj2/src/main/java/org/apache/qpid/protonj2/types/transport/ReceiverSettleMode.java:[25,25] [ImmutableEnumChecker] enums should be immutable: 'ReceiverSettleMode' has non-final field 'value'
    (see https://errorprone.info/bugpattern/ImmutableEnumChecker)
  Did you mean 'private final UnsignedByte value;'?
```

```
Warning:  /home/runner/work/qpid-protonj2/qpid-protonj2/protonj2/src/main/java/org/apache/qpid/protonj2/engine/impl/ProtonEndpoint.java:[63,17] [TypeParameterUnusedInFormals] Declaring a type parameter that is only used in the return type is a misuse of generics: operations on the type parameter are unchecked, it hides unsafe casts at invocations of the method, and it interacts badly with method overload resolution.
2219
    (see https://errorprone.info/bugpattern/TypeParameterUnusedInFormals)
```

**Comment by tabish121** [comment]
These tools can be useful when run from an IDE or something but I don't tend to find them useful in the maven build and especially errorprone which tends to break more often than provide useful results vs just a ton of warnings that don't actually point to any real issues that require constant babysitting

File: protonj2-client/src/test/java/org/apache/qpid/protonj2/client/util/ExternalMessage.java [hunk 1]
```diff
@@ -237,6 +237,7 @@ public String contentEncoding() {
     }
 
     @Override
+    @SuppressWarnings("NullTernary")
     public long absoluteExpiryTime() {
         return properties != null ? properties.getAbsoluteExpiryTime() : null;
     }
```
Comment (jiridanek) [inline protonj2-client/src/test/java/org/apache/qpid/protonj2/client/util/ExternalMessage.java:243]: This warning looks sensible to me. Returning `null` from a method with a primitive return type `long` is not possible.

File: protonj2/src/test/java/org/apache/qpid/protonj2/engine/impl/ProtonOutgoingDeliveryTest.java [hunk 2]
```diff
@@ -53,6 +54,7 @@ public void testSetGetMessageFormat() throws Exception {
         delivery.setMessageFormat(newFormat);
         assertEquals(newFormat, delivery.getMessageFormat(), "Unexpected message format");
 
+        // todo, that's 0, unlikely it is the highest value ;P
         // Highest value
         newFormat = (1 << 32) - 1;
```
Comment (jiridanek) [inline protonj2/src/test/java/org/apache/qpid/protonj2/engine/impl/ProtonOutgoingDeliveryTest.java:59]: Also sensible, `(1 << 32) - 1` there evaluates to 0. That is not what was intended by the test.

**Review by jiridanek — COMMENTED** [review]
I've put comments on the two things where I think the ERROR from Error Prone was justified. The rest is for things that would be weird in production code but make sense in tests. And then I suppressed the MissingSummary JavaDoc thing, which is related to javadoc codestyle more than to code correctness.

There are two `@SuppressWarnings` strings for some of them because the first is IntelliJ idea name for the check, and the second is Error Prone name for the same thing.

**Comment by jiridanek** [comment]
> These tools can be useful when run from an IDE or something but I don't tend to find them useful in the maven build and especially errorprone which tends to break more often than provide useful results vs just a ton of warnings that don't actually point to any real issues that require constant babysitting

Error Prone works quite well in an IDE. It emits compiler warnings, and IDEs are good at showing compiler warnings and making them clickable and navigable. So having the errorprone as an optional maven profile, with the nuisance noisy checks disabled in pom.xml can be reasonably convenient, assuming that there is an useful set of checks. If all checks are made to be warnings by default, and possibly only a select few of reliable checks are explicitly set as errors, that should minimize the amount of babysitting required. I consider this mostly a friendly feature for drive-by contributors: yay, some checker passed on my change! Of course, this requires that there is actually an errorprone checks that you care about enforcing.

Over all, I am currently most in favor of coverity, plus possibly running infer from time to time, manually.

**Pull request closed without merging**